### DISCUSSION TITLE
Device disabled

### DISCUSSION THREAD

**Discussion by Sebshino** [rustdesk/rustdesk-server-pro#182] [Q&A] [answered]
Hello,

I encounter a small problem, since the latest version of rustdesk pro, all rusdesk clients beyond the number of paid devices, it is deactivated....... I have the impression that the default group is taken in the count

Welcome to RustDesk Server Pro 1.1.13

Regards,
Seb

**Comment by rustdesk** [comment]
Proposed solutions:

- [ ] automatically delete/disable inactive clients 
- [x] custom RustDesk client which gives special id for such scenario which will not be taken into account  as managed devices
- [ ] remove this limit for some plan

BTW, turning on `Disable new devices `can also help you, you enable the device manually when you intend to support it.
![image](https://github.com/rustdesk/rustdesk-server-pro/assets/71636191/8c2f4e9a-4b7c-4854-9817-8a2dd5ffa621)

**Reply by Sebshino** [reply]
automatically delete inactive clients - > Ok
remove this limit for some plan -> Ok

or just a rusdesk quicksupport client (only ID and password) not integrating into the address book.
PS: quicksupport should not be incremented in our licenses
and keep the current thick client and the ingration in the address book

**Reply by cris-par** [reply]
Thanks for the replay:
automatically delete inactive clients - > Can be ok, but better if will be also possible delete manually unnecessary clients
custom RustDesk client... -> difficult to manage, because we don't know necessary at the beginning
remove this limit for some plans : OK

**Reply by mlopiccolo-progres** [reply]
For us, being able to have a custom client which just isn't added as a managed device would be perfect, it's how we worked back when we had anydesk. We are full of one time connections customers which dont need to be kept in the address book.

Enabling the Disable New Devices setting wouldnt work for us because we still need to add managed devices from time to time (servers).

Automatically deleting inactive clients would be a good compromise, but we would prefer to just have a custom client.

**Comment by rustdesk** [comment]
I have a solution for this, https://github.com/rustdesk/rustdesk/commit/7ea5a9bba35eeccbc19e061db29742171ff4a3c9
You can use this script to disable/delete devices yourself by `number of days offline`, `user name`, `device name`, `id`, `group name `etc.
It works since `RustDesk Server Pro` 1.3.1.

**Reply by rustdesk** [reply]
1. `RustDesk Server Pro` v1.3.1 required, https://github.com/rustdesk/rustdesk-server-pro/releases/tag/1.3.1
2. Go to `Settings` → `Tokens` → Create of web console and create a token with Device read/write permission.
![image](https://github.com/rustdesk/rustdesk-server-pro/assets/71636191/f8965af9-d23f-4557-b4b2-e5a8bbf05b47)
3. Download and run https://raw.githubusercontent.com/rustdesk/rustdesk/master/res/devices.py,
Supposing you want to delete devices which have been offline for more than 3 days.
```
# view all devices being offline 3 days first
python3 devices.py view --url http://rustdesk.yourcompany.com --token <token> --offline_days 3
# devices can be deleted only if they are disabled, so disable first
python3 devices.py disable --url http://rustdesk.yourcompany.com --token <token> --offline_days 3
python3 devices.py delete --url http://rustdesk.yourcompany.com --token <token> --offline_days 3
```
> Install `Python3` first, https://www.python.org/ and prerequisites of `devices.py`, `python3 -m pip install requests argparse`

> `--url` is identical to `API`, https://github.com/rustdesk/rustdesk/wiki/FAQ#failed-to-log-in-with-custom-client-build-or-public-release-with-wrong-api-settings

> You may see some `last_online': '2011-11-16T11:55:19`, it is not bug, it is the default value while migrating database to 1.3.1, the default value I use is my daughter's birthday.

> This script is an example, you are welcome to improve it.

**Reply by itxpress** [reply]
Had some time to play with this script a bit. It seems to be a bit wonky. So we have 80 devices listed in our admin console. But when I run this:

python3 devices.py view --url http://localhost:21114 --token [token] | wc -l 

I get 30, not 80. Ah OK - looked at script and the page size is 30, so you can only manipulate the first 30 devices. Our license is 100, so I changed the pageSize to 100 in the script and now we get all 80.

I'm guessing based on the answer edit above - the 2011-11 date is for really old devices that were pre migration. That's fine. But pre-migration to what? Pro? 1.3? 1.3.1? I think it's 1.3.1

Curious - I ran this to see what didn't have 2011-11 as last online

python3 devices.py view --url http://localhost:21114 --token [token] | grep -v "2011-11"

and the thing that sticks out is only devices with device_name populated are returned with a recent date. Most of our one and done devices don't have it because we weren't using the API in the past. Didn't need the address book. But when we found that the connection logs would not populate if you weren't logged in, we got it setup and now our techs are all logged into their Rustdesk clients. I'm guessing the information will be current from this point forward? So we're manually pruning the old stuff and hopefully this script will be useful going forward to keep the group managable... But can't say for certain if the lack of device data was not online since 1.3.1 upgrade or not connected to since we got our techs logged in via API.

But I would be careful using this to prune without manually checking devices first, because I *think* if you have devices that haven't been online since upgrading to 1.3.1, the last_online won't be populated so you could accidentally delete one you wanted to keep.

**Comment by Global-ti** [comment]
You can run the three commands at once.


```
import subprocess

def run_commands():
    commands = [
        'python devices.py view --url https://rust.url.com --token "apikey without quote" --offline_days 3',
        'python devices.py disable --url https://rust.url.com --token "apikey without quote" --offline_days 3',
        'python devices.py delete --url https://rust.url.com --token "apikey without quote" --offline_days 3'
    ]

    for command in commands:
        process = subprocess.run(command, shell=True)
        if process.returncode != 0:
            print(f"Command failed: {command}")
            break

if __name__ == "__main__":
    run_commands()
```

**Comment by rustdesk** [comment, marked as answer]
## What about casual support cases? We have some devices that need to be supported occasionally without unattended access.

*A new license model for this case: https://rustdesk.com/pricing#custom2*

[Excessive devices](https://github.com/rustdesk/rustdesk/wiki/FAQ#managed-devices) will not be disabled automatically; however, we have introduced a limit on the number of [concurrent connections](https://github.com/rustdesk/rustdesk/wiki/FAQ#concurrent-connections). Excessive devices cannot be assigned to users or device groups. Additionally, the maximum number of devices that can be included in a single address book is restricted by the total number of managed devices allowed under your license.

To prevent from casual support devices being added to devices page in web console, please check https://rustdesk.com/docs/en/self-host/client-configuration/advanced-settings/#register-device

**Reply by rustdesk** [reply]
You can also upgrade to this model in https://rustdesk.com/self-host/account

![image](https://github.com/user-attachments/assets/0eab4351-2883-49fe-8e1c-28d31244f6bb)

![image](https://github.com/user-attachments/assets/e1600dc5-e4db-4963-8070-737c068cb36d)

Your current value is 0 (unlimited), change it to 1 (same price as 0) or above, you will not get devices disabled any more by the system.

# Also do not forget to take it effect
You need to proceed to the `web console` to activate it manually as below. Just click on `Edit`, then `OK`, no need to edit anything, because your license key remains the same.

![image](https://github.com/user-attachments/assets/08f9f06c-aa47-49fe-b5be-9795b4baa77c)

You will see something below in the welcome page.

![image](https://github.com/user-attachments/assets/64cfe1a2-1ddf-4d92-aaf7-ab12c5178c31)
